File: phutball/game/game_logic/direction.py
from .location import Location

class Direction:
  '''Abstracts directions. Numeric coding is as on a standard number pad
  with 5 going nowhere.'''
  
  def __init__(self, num):
    self.delta_x = self.delta_y = 0
    if not (1 <= num <= 9):
      raise ValueError('Unknown direction code')

    # Delta x 
    if num in [1,4,7]:
      self.delta_x = -1
    elif num in [3,6,9]:
      self.delta_x = +1

    # Delta y
    if num in [1,2,3]:
      self.delta_y = -1
    elif num in [7,8,9]:
      self.delta_y = +1
      
    self.num = num
        
  def __add__(self, obj):
    try:
      return Location(
        (self.delta_x + obj.letter_index,
         self.delta_y + obj.number
        )
      )
    except ValueError: # Location does not exist
      return None
    
  def __radd__(self, obj):
    return self.__add__(obj)
  
  def __repr__(self):
    return '↙↓↘←.→↖↑↗'[self.num-1]

directions  = [Direction(n) for n in [1,2,3,4,6,7,8,9]]

File: react/phutball/src/gameLogic/direction.js
import Location from './location'

class Direction {
	constructor(num) {
		this.delta_x = 0;
		this.delta_y = 0;

		if ([1,4,7].includes(num)) {
			this.delta_x = -1
		} else if ([3, 6, 9].includes(num)) {
			this.delta_x = +1
		};

		if ([1,2,3].includes(num)) {
			this.delta_y = +1
		} else if ([7,8,9].includes(num)) {
			this.delta_y = -1
		}
	}

	add(loc) {
		return (Location.fromVector(
			loc.letterIndex + this.delta_y,
			loc.numberIndex + this.delta_x
		))
	}
}


// Should refactor to
// const directions = [1,2,3,4,6,7,8,9].map(num => new Direction(num))
var directions = Array(8);

for (var i = 0; i < 8; i ++) {
	directions[i] = new Direction([1,2,3,4,6,7,8,9][i])
}

export default directions
